quickwit-cluster/docs/kubernetes-discovery.md

```diff
@@ -0,0 +1,195 @@
+deployments and statefulsets), but we will consider this potential topology (inspired from Bitnami's
+[Cassandra chart](https://github.com/bitnami/charts/tree/master/bitnami/cassandra/templates) and my experience)
+as an illustration for discussions on the proposed solution:
+
+- a `$release-indexing` deployment for an elastic pool of indexer nodes, that can be quickly auto-scaled based on volume
+- a `$release-indexing` service, to provide a DNS endpoint for the push API. This service uses a label selector to 
+  expose pods of the `$release-indexing` deployment once [they are ready](https://blog.colinbreck.com/kubernetes-liveness-and-readiness-probes-how-to-avoid-shooting-yourself-in-the-foot/)
+- a `$release-search` statefulset, for a pool of searcher nodes. Statefulset would allow prioritizing disk caching
+  (with persistent disk volumes) over elasticity (they are slower to up/down scale), if the read volume is steady
+- a `$release-search` service, to provide a DNS endpoint for the search API. It uses a label selector to
+  expose ready pods of the `$release-search` statefulset
+- a `$release-headless` service, to be used for discovery (see next section). It will be set:
+  - as [headless](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services), to allow
+    enumerating all pods at the DNS level
+  - with `publishNotReadyAddresses: true` to expose all pods, even before they are ready
+  - with a label selector matching both the `$release-indexing` and `$release-search` pods, as a unified cluster
+
+## Proposed solution
+
+The following action items are ordered by priority, and could be independently implemented in several releases.
+Alternatives are discussed in the next section, with a pros/cons comparison.
+
+### 1a. Source the sibling pods as seeds via DNS enumeration on startup
+
+A new `peer_seed_service` configuration option is added to Quickwit. It is exclusive to the existing `peer_seeds`,
+and setting both will result on a validation error. `quickwit-cluster::start_cluster_service` is updated to check
+this option. If it is not empty:
+
+- a DNS enumeration request is issued, that will return the IP addresses of all the live pods in the cluster,
+- if the DNS request fails or if the result is empty, Quickwit fails to start. Kubernetes will restart the
+  container, allowing to retry the request when more pods are created. In case of repeated failures (misconfiguration,
+  or DNS error), the CrashLoopBackoff mechanism will engage, reducing the pressure on the DNS server by keeping
+  pods down longer between retries,
+- if not empty, the response is used to populate the `seed_nodes` argument passed to `scuttlebutt`.
+
+Due to the eventually consistent nature of the k8s control plane, there will be a delay before all pods are exposed
+in the DNS record. We do have guarantees that the responses to the different pods will be consistent. Pods will
```

Comment:
If we have pod A and pod B that starts at the same time, can you confirm that it is impossible for k8s DNS to deliver "eventually consistent results in the sense of ":

- node A receives only node A IP
- node B receives only node B IP.

Hence creating a split brain.

Reply:
The Kubernetes control plane is implemented as a series of single-threaded loops (called controllers) to avoid race conditions. This ensures that pods will be registered in the [Endpoints](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#endpoints-v1-core) entity in a consistent order.

That `Endpoints` entity is the source of truth for the DNS record. We might encounter a pathological DNS layer with inconsistent responses, but the common expectation is for the record to fully reflect one version of the `Endpoints` entity (not necessarily the latest though, due to propagation).

If A is registered in the `Endpoints` entity before B, we should  get either A, or A+B, never B alone.

Reply:
And we are certain the DNS will always serve a view consistent with a past state of the control plane?

Could there be a silly situation where a cloud provider uses an adhoc distributed DNS caching service that uses a replication with eventual consistency semantics that could yield this situation?
(no sure if this is a dumb question or not.)

Reply:
Hmmm there a lot DNS caching solutions... But they all act ... as a cache. I'm overthinking this.

Reply:
The k8s ecosystem is so fragmented (especially in the networking part), that it is difficult to give a definitive answer on that. My anecdotal experience with the two most popular solutions ([kube-dns](https://github.com/kubernetes/dns) and [coredns](https://github.com/coredns/coredns/tree/master/plugin/kubernetes)) makes me pretty confident on this solution, but users could run alternative DNS servers with a buggy implementation.

Other solutions make this assumptions, so I would expect the risk for this to be undetected to be low. If that were to happen, there are easy mitigations we can suggest:

- deploy new clusters with a single pod (`replicas: 1`), then manually upscale once the DNS entry is created,
- add an entrypoint or an init container delaying the start of the Quickwit process for a couple of seconds, to ensure the DNS has settled.